File: samples/perfget.c
// Simple performance test of C hash table get

// See perftest.sh for results

#include "../ht.h"

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void exit_nomem(void) {
    fprintf(stderr, "out of memory\n");
    exit(1);
}

void* found;

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: perftest file\n");
        return 1;
    }

    // Read entire file into memory.
    FILE* f = fopen(argv[1], "rb");
    if (f == NULL) {
        fprintf(stderr, "can't open file: %s\n", argv[1]);
        return 1;
    }
    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    fseek(f, 0, SEEK_SET);
    char* contents = (char*)malloc(size + 1);
    if (contents == NULL) {
        exit_nomem();
    }
    size_t nread = fread(contents, 1, size, f);
    if ((long)nread != size) {
        fprintf(stderr, "read %ld bytes instead of %ld", (long)nread, size);
        return 1;
    }
    fclose(f);
    contents[size] = 0;

    ht* counts = ht_create();
    if (counts == NULL) {
        exit_nomem();
    }

    for (char* p = contents; *p;) {
        // Skip whitespace.
        while (*p && *p <= ' ') {
            p++;
        }
        char* word = p;

        // Find end of word.
        while (*p && *p > ' ') {
            p++;
        }
        if (*p != 0) {
            *p = 0;
            p++;
        }

        // Look up word.
        void* value = ht_get(counts, word);
        if (value != NULL) {
            // Already exists, increment int that value points to.
            int* pcount = (int*)value;
            (*pcount)++;
            continue;
        }

        // Word not found, allocate space for new int and set to 1.
        int* pcount = malloc(sizeof(int));
        if (pcount == NULL) {
            exit_nomem();
        }
        *pcount = 1;
        if (ht_set(counts, word, pcount) == NULL) {
            exit_nomem();
        }
    }

    // Copy keys to array
    const char** keys = malloc(ht_length(counts) * sizeof(char*));
    if (keys == NULL) {
        exit_nomem();
    }
    hti it = ht_iterator(counts);
    int i = 0;
    while (ht_next(&it)) {
        keys[i] = it.key;
        i++;
    }

    int runs = 10;
    clock_t start = clock();
    for (int run=0; run<runs; run++) {
        for (int i=0; i<ht_length(counts); i++) {
            found = ht_get(counts, keys[i]);
        }
    }
    clock_t end = clock();
    double elapsed_ms = (double)(end - start) / CLOCKS_PER_SEC * 1000;
    printf("%d runs getting %lu keys: %.03fms\n", runs, ht_length(counts), elapsed_ms);

    return 0;
}

File: samples/perfget.go
// Simple performance test of Go map get

package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"time"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "usage: perftest file\n")
		os.Exit(1)
	}

	contents, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "error reading file: %v\n", err)
		os.Exit(1)
	}

	counts := make(map[string]int)
	for i := 0; i < len(contents); {
		// Skip whitespace.
		for i < len(contents) && contents[i] <= ' ' {
			i++
		}
		start := i

		// Find end of word.
		for i < len(contents) && contents[i] > ' ' {
			i++
		}
		if i > start {
			word := contents[start:i]
			counts[string(word)]++
		}
	}

	// Copy keys to slice
	keys := make([]string, 0, len(counts))
	for k := range counts {
		keys = append(keys, k)
	}

	const runs = 10
	start := time.Now()
	for run := 0; run < runs; run++ {
		for i := 0; i < len(counts); i++ {
			_ = counts[keys[i]]
		}
	}
	end := time.Now()
	elapsed := end.Sub(start)
	fmt.Printf("%d runs getting %d keys: %v\n", runs, len(counts), elapsed)
}
